[RISC-V] Clean up NYI_RISCV64 in CodeGen and Emitter

yurai007 opened the pull request:
This PR focus mostly on NYI_RISCV64s that are easy to fix, remove or classify as ones that will need more care in future.

Part of https://github.com/dotnet/runtime/issues/84834
cc @wscho77 @HJLeee @clamp03 @JongHeonChoi @t-mustafin @gbalykov @ashaurtaev @sirntar @tomeksowi @Bajtazar

yurai007 committed:
[RISC-V] Remove redundant GT_NEG/GT_NOT switch cases

In 6ca0784d82e4 commit, function genCodeForNegNot diverged from her LoongArch friend
and didn't call genGetInsForOper since then. Therefore we can remove
GT_NEG/GT_NOT integer cases and GT_NEG float case as there are all dead.

yurai007 committed:
[RISC-V] Remove redundant genCodeForLoadOffset definition

Given that such function doesn't make sense outside x86/x64 mov instruction context
we can get rid of definition.

yurai007 committed:
[RISC-V] Remove redundant emitInsToJumpKind definition

It looks like emitInsToJumpKind is helper function coming from ARM large jump handling logic.
Currently RISC-V emitter doesn't rely on instruction format
and probably won't need anything like ARM large jump handling.

yurai007 committed:
[RISC-V] Replace NYI_RISCV64 with NO_WAY in switch statements

For the cases when unreachability is easy to prove we can
express that fact by using proper NO_WAY macro.

yurai007 committed:
[RISC-V] Clean up emitInsTernary function from NYI_RISCV64

When control flow reach needCheckOv in intConst block,
only addi/addiw instructions can be handled there and first NYI_RISCV64 is redundant.
Remaining condition statements starting with GT_MUL case can be replaced
with more straightforward switch statement and then second NYI_RISCV64 is redundant.

yurai007 committed:
[RISC-V] Update some NYI_RISCV64 with comments

yurai007 commented on src/coreclr/jit/emitriscv64.cpp:
```diff
@@ -4207,272 +4207,276 @@ regNumber emitter::emitInsTernary(instruction ins, emitAttr attr, GenTree* dst,
         {
-            if (ins == INS_addi || ins == INS_addiw)
+            // At this point andi/ori/xori are excluded by previous checks
+            assert(ins == INS_addi || ins == INS_addiw);
```
Diff starting from here is terrible due to indentation changes. Commit message highlights what is really going on here.

tomeksowi replied:
"Hide whitespaces" in Diff view settings helps with that.

yurai007 committed:
[RISC-V] Implement remaining NYI_RISCV64 in genJumpToThrowHlpBlk_la

When no helper for codeKind is available we need to emit
appropriate code directly.

tomeksowi commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -4020,6 +4020,22 @@ int CodeGenInterface::genCallerSPtoInitialSPdelta() const
 }
 
+// Produce generic and unoptimized code for loading constant to register
+static void emitLoadImmediateDirectly(emitter* emit, regNumber dstRegister, ssize_t imm)
+{
+    ssize_t high = (imm >> 32) & 0xffffffff;
+    emit->emitIns_R_I(INS_lui, EA_PTRSIZE, dstRegister, (((high + 0x800) >> 12) & 0xfffff));
+    emit->emitIns_R_R_I(INS_addi, EA_PTRSIZE, dstRegister, dstRegister, (high & 0xfff));
+
+    ssize_t low = imm & 0xffffffff;
+    emit->emitIns_R_R_I(INS_slli, EA_PTRSIZE, dstRegister, dstRegister, 11);
+    emit->emitIns_R_R_I(INS_addi, EA_PTRSIZE, dstRegister, dstRegister, ((low >> 21) & 0x7ff));
+
+    emit->emitIns_R_R_I(INS_slli, EA_PTRSIZE, dstRegister, dstRegister, 11);
+    emit->emitIns_R_R_I(INS_addi, EA_PTRSIZE, dstRegister, dstRegister, ((low >> 10) & 0x7ff));
+    emit->emitIns_R_R_I(INS_ld, EA_PTRSIZE, dstRegister, dstRegister, (low & 0x3ff));
+}
```
Nit: the name suggests it's an implementation of `li` pseudo or a emitter::emitLoadImmediate alternative but it does a deref at the end with low 10 bits as the offset.

tomeksowi commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -7180,9 +7186,24 @@ inline void CodeGen::genJumpToThrowHlpBlk_la(
+                // TODO: In the future we may consider using emitter::emitLoadImmediate instead,
+                // which is less straightforward but offers slightly better codegen.
+                emitLoadImmediateDirectly(GetEmitter(), callTarget, (ssize_t)pAddr);
+                emit->emitIns_R_R_I(INS_ld, EA_PTRSIZE, callTarget, callTarget, 0);
```
emitLoadImmediateDirectly already does a deref with a ld, you need one more?

tomeksowi commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -7210,9 +7186,24 @@ inline void CodeGen::genJumpToThrowHlpBlk_la(
+                // TODO: In the future we may consider using emitter::emitLoadImmediate instead,
+                // which is less straightforward but offers slightly better codegen.
+                emitLoadImmediateDirectly(GetEmitter(), callTarget, (ssize_t)pAddr);
+                emit->emitIns_R_R_I(INS_ld, EA_PTRSIZE, callTarget, callTarget, 0);
```
emitLoadImmediateDirectly already derefs at the end, you need one more?

yurai007 replied:
Good catch, I think it's wrong, but let me check emitted assembly first.

clamp03 commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -4042,6 +4020,23 @@ int CodeGenInterface::genCallerSPtoInitialSPdelta() const
     return callerSPtoSPdelta;
 }
 
+// Produce generic and unoptimized code for loading constant to register and dereferencing it
+// at the end
+static void emitLoadConstAtAddr(emitter* emit, regNumber dstRegister, ssize_t imm)
+{
+    ssize_t high = (imm >> 32) & 0xffffffff;
+    emit->emitIns_R_I(INS_lui, EA_PTRSIZE, dstRegister, (((high + 0x800) >> 12) & 0xfffff));
+    emit->emitIns_R_R_I(INS_addi, EA_PTRSIZE, dstRegister, dstRegister, (high & 0xfff));
+
+    ssize_t low = imm & 0xffffffff;
+    emit->emitIns_R_R_I(INS_slli, EA_PTRSIZE, dstRegister, dstRegister, 11);
+    emit->emitIns_R_R_I(INS_addi, EA_PTRSIZE, dstRegister, dstRegister, ((low >> 21) & 0x7ff));
+
+    emit->emitIns_R_R_I(INS_slli, EA_PTRSIZE, dstRegister, dstRegister, 11);
+    emit->emitIns_R_R_I(INS_addi, EA_PTRSIZE, dstRegister, dstRegister, ((low >> 10) & 0x7ff));
+    emit->emitIns_R_R_I(INS_ld, EA_PTRSIZE, dstRegister, dstRegister, (low & 0x3ff));
+}
+
```
As I know, the reason why this unoptimized code is used (instead of `emitLoadImmediate`) is there are some fixed number for code size. Am I right? If so, I hope you keep this in an optimization to-do list for later. Thank you!

yurai007 replied:
Exactly, that's the reason. I already left _TODO_ comment on caller side: [link](https://github.com/dotnet/runtime/pull/94666/commits/a0c425645c7bf6ffdc449788351df19ed9d4ba55#diff-00a478e27ec7f1116fa2dc1a5976b1a4f6762565dc92af9526d1af11958e9a94R7202). Adapting _emitLoadImmediate_ for our purpose shouldn't be a problem, but yeah, I considered it as codegen optimization, and left it for later.

clamp03 replied:
Please update TODO format based on https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/clr-jit-coding-conventions.md#7.1.5. I think `TODO:` should be `TODO-RISCV64-CQ`.

clamp03 commented:
@jakobbotsch Could you review this PR? Thank you.
